### PR TITLE
Allow function and environment creation on all namespaces

### LINKED ISSUE #2822: Environment pods do not show up when environment is created in another namespace
**Fission/Kubernetes version**

<pre>
$ fission version

client/server:
  fission/core:
    Version: v1.19.0

$ kubectl version

Client Version: v1.27.2
Kustomize Version: v5.0.1
Server Version: v1.27.3

</pre>

**Kubernetes platform (e.g. Google Kubernetes Engine)**

kind v0.20.0 (under MacOS)

**Describe the bug**

When I create a namespace test, and run:

> fission env create --name nodejs --image fission/node-env -n test

I don't see any pods showing up.

The nodejs environment is created though:

<pre>

NAME   IMAGE            BUILDER_IMAGE POOLSIZE MINCPU MAXCPU MINMEMORY MAXMEMORY EXTNET GRACETIME NAMESPACE
nodejs fission/node-env               3        0      0      0         0         false  0         test

</pre>

<br>

I experimented with different sort of settings:
* by setting the NAMESPACE environment variables (using the export command)
* by updating the NAMESPACE settings in values.yaml
* by using: kubectl config set-context --current  --namespace=test

but none of that seems to have effect.

<br>
    
I also tried Microk8s instead of kind, but I get the same results.

<br>

**To Reproduce**

> kubectl create namespace test
> fission env create --name nodejs --image fission/node-env -n test
> kubect get pods -A

**Expected result**

When I follow the standard nodejs example, and run:

> fission env create --name nodejs --image fission/node-env
> kubect get pods -A

I see the following environment pods:

default       poolmgr-nodejs-default-5261-6cc484b9b7-5xsfd   2/2     Running   0          72s
default       poolmgr-nodejs-default-5261-6cc484b9b7-j4c85   2/2     Running   0          72s
default       poolmgr-nodejs-default-5261-6cc484b9b7-52sqw   2/2     Running   0          72s

<br>

I expect the same result when I create the environment in a namespace test:

test       poolmgr-nodejs-default-5261-6cc484b9b7-5xsfd   2/2     Running   0          72s
test       poolmgr-nodejs-default-5261-6cc484b9b7-j4c85   2/2     Running   0          72s
test       poolmgr-nodejs-default-5261-6cc484b9b7-52sqw   2/2     Running   0          72s

**Actual result**

I don't see any pods showing up.

### PR DESCRIPTION
**Pull request by fragsalat**
## Description
Related to #2822 I tried to allow function and environment creation on any namespace. In our setup teams use namespaces per application and namespaces are created over time, which is just not doable to maintain them in a config. Please bear with me if something looks odd, this is my first time touching go and k8s apis. I added another setting `fissionOnAllNamespaces` which will , if true, create all Roles and RoleBindings as ClusterRole and ClusterRolebinding and only create them in the default namespace and not in individual function, builder or additionalNamespaces. Additionally the NamespaceResolver is returning `metav1.NamespaceAll`, which is equal to `""` instead of individual namespaces. This leads to listers and listeners to be created for the global scope instead of single namespaces. Last but not least I moved the service account check before creating a generic pool or a new deployment to ensure the permission exist on the namespace, as this can not be guaranteed while installing fission. For this to work, at the moment the `executor.serviceAccountCheck.enabled` flag needs to be true.

I am looking forward to read your thoughts on that approach and for your improvement ideas. As I am new to Go I did not add any new test and hope you may can support me here.

## Which issue(s) this PR fixes:
Fixes #2822

## Testing
- Installed on minikube with `fissionOnAllNamespaces` and `executor.serviceAccountCheck.enabled` set to true
- Created environments on default and custom namespaces like `test` with poolsize 0, 1+
- Created functions on default and custom namespaces like `test` with default executortype and newdeploy
- Used fission cli to test the function response

## Checklist:

- [x] I ran tests as well as code linting locally to verify my changes. 
- [x] I have done manual verification of my changes, changes working as expected.
- [ ] I have added new tests to cover my changes.
- [ ] My changes follow contributing guidelines of Fission.
- [ ] I have signed all of my commits.

I plan to check the rest of the checklist once I got your input on the feasibility of this :)
Have a great day and week

### CODE REVIEW

**Comment by fragsalat** [comment]
Good morning. I was wondering whether someone found time to look at this idea? I would like to get feedback on that before investing further time to polish the PR :)

File: charts/fission-all/values.yaml [hunk 1]
```diff
@@ -191,7 +195,7 @@ executor:
 
   serviceAccountCheck:
     ## enables fission to create service account, roles and rolebinding for missing permission for builder and fetcher.
-    enabled: true
+    enabled: false
```
Comment (sanketsudake) [inline charts/fission-all/values.yaml:198]: Please keep default values intact.

File: skaffold.yaml [hunk 2]
```diff
@@ -49,8 +49,10 @@ manifests:
         imageTag: ""
         influxdb.enabled: "false"
         namespace: fission
-        builderNamespace: ""
-        functionNamespace: ""
+        builderNamespace: "builder"
+        functionNamespace: "function"
+        additionalNamespaces: ["test"]
+        fissionOnAllNamespaces: "true"
```
[the remark this answers is not in the document]
Comment (fragsalat) [reply skaffold.yaml]: Reverted 👍

File: pkg/utils/serviceaccount_test.go [hunk 3]
```diff
@@ -21,7 +21,7 @@ func TestServiceAccountCheck(t *testing.T) {
 	kubernetesClient := fake.NewSimpleClientset()
 	logger := loggerfactory.GetLogger()
 	os.Setenv(ENV_CREATE_SA, "true")
-	CreateMissingPermissionForSA(ctx, kubernetesClient, logger)
+	CreateMissingPermissionForSA(ctx, kubernetesClient, logger, "default")
```
[the remark this answers is not in the document]
Comment (fragsalat) [reply pkg/utils/serviceaccount_test.go]: changed 👍

File: charts/fission-all/values.yaml [hunk 4]
```diff
@@ -67,20 +67,24 @@ defaultNamespace: default
 ## if builderNamespace is set to empty then builder resources will be created in the same namespace as the Fission resources.
 ## This is different from the release namespace.
 ##
-builderNamespace: ""
+builderNamespace: "builder"
```
[the remark this answers is not in the document]
Comment (fragsalat) [reply charts/fission-all/values.yaml]: Reverted 👍

File: charts/fission-all/values.yaml [hunk 5]
```diff
@@ -67,20 +67,24 @@ defaultNamespace: default
...
 
 ## functionNamespace represents the namespace in which Fission Function resources will be created.
 ## if functionNamespace is set to empty then function resources will be created in the same namespace as the Fission resources.
 ## This is different from the release namespace.
 ##
-functionNamespace: ""
+functionNamespace: "functions"
```
[the remark this answers is not in the document]
Comment (fragsalat) [reply charts/fission-all/values.yaml]: Reverted 👍

File: charts/fission-all/values.yaml [hunk 6]
```diff
@@ -67,20 +67,24 @@ defaultNamespace: default
...
 
 ## Fission will watch the following namespaces along with the `defaultNamespace` for fission custom resources.
 ## additionalFissionNamespaces: 
 ## - namespace1
 ## - namespace2
 ## - namespace3
-additionalFissionNamespaces: []
+additionalFissionNamespaces: ["test"]
```
[the remark this answers is not in the document]
Comment (fragsalat) [reply charts/fission-all/values.yaml]: Reverted 👍

File: charts/fission-all/values.yaml [hunk 7]
```diff
@@ -67,20 +67,24 @@ defaultNamespace: default
...
+
+## Allows to create functions and environments on all namespaces in the cluster. This overwrites additionalFissionNamespaces setting
+## Setting this requires executor.serviceAccountCheck.enabled to be true in order to create service accounts and permissions when creating environments and functions
+fissionOnAllNamespaces: true
```
[the remark this answers is not in the document]
Comment (fragsalat) [reply charts/fission-all/values.yaml]: Changed it to false, are you adding the warining or should I do? If I have to do it, where should I add it?

File: charts/fission-all/values.yaml (hunk 1 above)
Comment (fragsalat) [reply charts/fission-all/values.yaml]: Reverted 👍

**Comment by fragsalat** [comment]
I guess as you just commented on default values you are fine with the overall Idea of the solution? :)

**Comment by fragsalat** [comment]
I've setup a cluster locally using kind and executed the steps in the workflow manually and the failing function `test_fission_objects` did not fail for me. So idk what to do to reproduce and fix this failing test.

**Comment by sanketsudake** [comment]
"Fission CI upgrade" tests fail sometimes and failure is intermittent. You can ignore the failure for now, needs bit of investigation

**Review by sanketsudake — COMMENTED** [review]
Overall, we can follow two approaches here.

1. Creating cluster roles and role binding  upfront as we have done in the PR
2. Creating role and role binding for each namespace via the executor

In 1st case, the user is aware that they are enabling Fission across multiple namespaces and cleanup would be managed by Helm chart on uninstallation of fission.
In 2nd case, we would be permitting the executor to create a role and role binding to the executor pod which could be more of a security problem. Also cleanup needs to be managed without a helm chart here.

 So the approach in this PR seems better. It would be good to have an integration test flow separate from existing integration tests and clone of existing integration tests to ensure this solution is working e2e.

File: charts/fission-all/templates/_fission-kuberntes-role-generator.tpl [hunk 8]
```diff
@@ -1,7 +1,7 @@
 {{- define "kubernetes-role-generator" }}
 ---
 apiVersion: rbac.authorization.k8s.io/v1
-kind: Role
+kind: {{ if .Values.fissionOnAllNamespaces }}Cluster{{- end }}Role
```
[the remark this answers is not in the document]
Comment (fragsalat) [reply charts/fission-all/templates/_fission-kuberntes-role-generator.tpl:4]: Can do so. Funny enough I had that in first place and found it pollutes the templates and is harder to read which is why I changed it to this format :smile:

**Comment by fragsalat** [comment]
> Overall, we can follow two approaches here.
> 
> 1. Creating cluster roles and role binding  upfront as we have done in the PR
> 2. Creating role and role binding for each namespace via the executor
> 
> In 1st case, the user is aware that they are enabling Fission across multiple namespaces and cleanup would be managed by Helm chart on uninstallation of fission. In 2nd case, we would be permitting the executor to create a role and role binding to the executor pod which could be more of a security problem. Also cleanup needs to be managed without a helm chart here.
> 
> So the approach in this PR seems better. It would be good to have an integration test flow separate from existing integration tests and clone of existing integration tests to ensure this solution is working e2e.

I want to note that with the first approach, which I implemented, it is still necessary to dynamically create he role binding for the service accounts in the executor. The namespace the function or env is deployed to might not have existed at the time of installation. That makes the serviceAccountChecke to be enabled if fissionOnAllNamespaces is true. Maybe the fissionOnAllNamespace in that case should overwrite the `executor.serviceAccountCheck.enabled` in case it is false but fissionOnAllNamespaces is true. Wdyt?

**Comment by fragsalat** [comment]
> So the approach in this PR seems better. It would be good to have an integration test flow separate from existing integration tests and clone of existing integration tests to ensure this solution is working e2e.

I am a bit uncertain about what you consider as integration test here and I yet could not find any technical documentation about the project. Do you mean the test/e2e/* as integration test? I could not find out where and how they are actually used. I also could not find out where most of the files in test/tests* are used like test_ingress.sh. Can you please give me some guideline where to start in regards to the integration tests?

**Comment by sanketsudake** [comment]
Thinking about use-cases mentioned in also https://github.com/fission/fission/issues/2913
We may need to support wildcard namespaces

**Comment by fragsalat** [comment]
So I created now a new flow in the pr template and adjusted the previously disabled function test to create a valid env and  function either in standard namespace or in custom namespace depending on the flow. Is that enough or would you expect more to test?
Btw, Running all tests locally did not succeed properly so I hope it works here^^

**Comment by fragsalat** [comment]
@sanketsudake could you maybe run the workflow so I can see it passes or not? Is there a stable way to run this locally?

**Comment by fragsalat** [comment]
I think I fixed the existing integration test but I don't know why the all-ns tests time out. I try running them locally with act but I get multiple test failures for several reasons. Will take me a while to reconstruct the result in github. Is there a way to see the log file contents even when there is a timeout? @sanketsudake

**Comment by fragsalat** [comment]
@sanketsudake could you please re-run the workflow? At least I expect the original workflow to pass. As mentioned its hard to test these locally, do you have any suggestion how to move forward?

**Comment by dsegurakaliop** [comment]
Hi !

We need this feature a lot. 

We are available to test it in cluster env directly in our labs if necessary.

Can we expect to reopen this PR ?

Best regards,

Didier

**Pull request closed without merging**